src/Std/Data/DHashMap/Lemmas.lean

```diff
@@ -964,4 +965,701 @@ theorem distinct_keys [EquivBEq α] [LawfulHashable α] :
+      (cast (congrArg (Option ∘ β) (eq_of_beq h)) (f (m.get? k)))
+    else m.get? k' :=
+  sorry
+
+
+@[simp]
+theorem get?_alter_self [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).get? k = f (m.get? k) :=
+  sorry
+
+theorem get_alter [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)}
+    (h : k' ∈ m.alter k f) :
+    (m.alter k f).get k' h =
+    if heq : k == k' then
+      haveI h' : (f (m.get? k)).isSome := by rwa [mem_alter, if_pos heq] at h
+      cast (congrArg β (eq_of_beq heq)) <| (f (m.get? k)).get h'
+    else
+      haveI h' : k' ∈ m := by rwa [mem_alter, if_neg heq] at h
+      m.get k' h' :=
+  sorry
+
+@[simp]
+theorem get_alter_self [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)}
+    {h : k ∈ m.alter k f} :
+    haveI h' : (f (m.get? k)).isSome := by rwa [mem_alter_self] at h
+    (m.alter k f).get k h = (f (m.get? k)).get h' :=
+  sorry
+
+theorem get!_alter [LawfulBEq α] {k k' : α} [hi : Inhabited (β k')] {f : Option (β k) → Option (β k)} :
+    (m.alter k f).get! k' =
+    if heq : k == k' then
+      haveI : Inhabited (β k) := ⟨cast (congrArg β <| eq_of_beq heq).symm default⟩
+      cast (congrArg β (eq_of_beq heq)) <| (f (m.get? k)).get!
+    else
+      m.get! k' :=
+  sorry
+
+@[simp]
+theorem get!_alter_self [LawfulBEq α] {k : α} [Inhabited (β k)] {f : Option (β k) → Option (β k)} :
+    (m.alter k f).get! k = (f (m.get? k)).get! :=
+  sorry
+
+theorem getD_alter [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).getD k' =
+    if heq : k == k' then
+      f (m.get? k) |>.map (cast (congrArg β <| eq_of_beq heq)) |>.getD
+    else
+      m.getD k' :=
+  sorry
+
+@[simp]
+theorem getD_alter_self [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).getD k = (f (m.get? k)).getD :=
+  sorry
+
+theorem getKey?_alter [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).getKey? k' =
+    if k == k' then
+      (f (m.get? k)).map (fun _ => k)
+    else
+      m.getKey? k' :=
+  sorry
+
+@[simp]
+theorem getKey?_alter_self [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).getKey? k = (f (m.get? k)).map (fun _ => k) :=
+  sorry
+
+theorem getKey!_alter [LawfulBEq α] [Inhabited α] {k k' : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).getKey! k' =
+    if k == k' then
+      (f (m.get? k)).map (fun _ => k) |>.get!
+    else
+      m.getKey! k' :=
+  sorry
+
+@[simp]
+theorem getKey!_alter_self [LawfulBEq α] [Inhabited α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).getKey! k = ((f (m.get? k)).map (fun _ => k)).get! :=
+  sorry
+
+theorem getKey_alter [LawfulBEq α] [Inhabited α] {k k' : α} {f : Option (β k) → Option (β k)}
+    (h : k' ∈ m.alter k f) :
+    (m.alter k f).getKey k' h =
+    if heq : k == k' then
+      k
+    else
+      haveI h' : k' ∈ m := by rwa [mem_alter, if_neg heq] at h
+      m.getKey k' h' :=
```

Comment:
I think this is out of scope for initial work, but I wonder if an autoparam for the proofs in `get` and `getKey`, backed by a simple tactic, would work. I suspect we could remove a lot of these inline proofs with a custom tactic for proving membership.

Reply:
Theoretically, `simpa` -- with suitable arguments -- is indeed able to finish most of these instances. However, perhaps half of the ugly `haveI`'s aren't actually about passing a lemma to `DHashMap.getKey/get`, but about passing a lemma to `Option.get`. I think it would be relatively simple to create tactics for both of these cases, but the latter case seems to me to be overly specific and perhaps not worth a tactic?

Either way, I won't do this in this PR, but I'm very interested in your thoughts and might do it in another PR!